### PR TITLE
Add customLegendTypes prop to customise covered legend types

### PR DESCRIPTION
**Pull request by Clebal**
Added `customLegendTypes` prop in LegendWidget and LegendWidgetUI to customise legend types.

Example:

```javascript
const CUSTOM_LEGEND_TYPES = {
      'my-custom-legend': MyCustomLegend
}

function App {
      return <LegendWidget customLegendTypes={CUSTOM_LEGEND_TYPES} />
}

function MyCustomLegend () {
      return <p>Test</p>
}

//

dispatch(
  updateLayer({
    id: '...',
    layerAttributes: {
      legend: {
        type: 'my-custom-legend',
        ...
      }
    }
  })
)
```

Result:

image

### CODE REVIEW

**Comment by Clebal** [comment]
After talking with @borja-munoz, we decided to use `customLegendTypes` to allow to only add custom legend types instead of having the possibility to add and modify the current ones. When the legend type isn't one of the covered by default, the LegendWidgetUI checks the `customLegendTypes` object, if it doesn't exists, it renders an error message.

image

**Review by zbigg — COMMENTED** [review]
LGTM, not sure if we need more approvals

**Comment by Clebal** [comment]
@zbigg yes, we need one more approval since this is an enhancement. Btw, you commented, not approved.

**Review by zbigg — APPROVED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,8 @@
 - Fix legends typings [#335](https://github.com/CartoDB/carto-react/pull/335)
 - Fix legends export [#334](https://github.com/CartoDB/carto-react/pull/334)
 - Fix duplicated logic for category selection in PieWidgetUI [#332](https://github.com/CartoDB/carto-react/pull/332)
+- Add customLegendTypes prop to customise covered legend types [#325](https://github.com/CartoDB/carto-react/pull/325)
+- Add error label if legend type is unknown [#325](https://github.com/CartoDB/carto-react/pull/325)
```
Comment (VictorVelarde) [inline CHANGELOG.md:10]: I think we don't need this entry, as presenting error label is part or the implementation

File: packages/react-ui/src/widgets/legend/LegendWidgetUI.js [hunk 2]
```diff
@@ -74,11 +76,13 @@ function LegendWidgetUI({
 
 LegendWidgetUI.defaultProps = {
   layers: [],
+  customLegendTypes: {},
   collapsed: false
 };
 
 LegendWidgetUI.propTypes = {
   className: PropTypes.string,
+  customLegendTypes: PropTypes.objectOf(PropTypes.func),
```
Comment (VictorVelarde) [inline packages/react-ui/src/widgets/legend/LegendWidgetUI.js:85]: Maybe use PropTypes.shape with exact props ?

File: packages/react-ui/src/widgets/legend/LegendWidgetUI.js [hunk 3]
```diff
@@ -187,23 +197,22 @@ function LegendRows({ layers = [], onChangeVisibility, onChangeOpacity }) {
...
             note = '',
             attr = ''
           } = legend;
 
           const isLast = layers.length - 1 === index;
-          // TODO: Add validation for layer.type
-          const hasChildren = LEGEND_COMPONENT_BY_TYPE[type] || children;
-          const LegendComponent = LEGEND_COMPONENT_BY_TYPE[type] || (() => children);
+          const LegendComponent =
+            LEGEND_COMPONENT_BY_TYPE[type] || customLegendTypes[type] || UnknownLegend;
```
Comment (VictorVelarde) [inline packages/react-ui/src/widgets/legend/LegendWidgetUI.js:208]: Ok, makes sense to keep existing ones, 👍🏻

**Review by VictorVelarde — APPROVED** [review]
LGTM

File: CHANGELOG.md (hunk 1 above)
Comment (Clebal) [reply CHANGELOG.md:10]: Ok, I'll remove it

**Pull request merged**